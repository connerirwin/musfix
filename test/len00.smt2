; Polymorphic parameters are currently replaced with ints

; (qualif ListZ ((v [@0]) (>= (len v) 0)))
(qualif ListZ ((v [Int])) (>= (len v) 0))
; (qualif ListZ ((v [a])) (>= (len v) 0))

; uninterpreted function len that takes 2 polymorphic parameters,
; @(0) and @(1) and returns an Int

; number is the # of polymorphic params
; TODO why is it explicitly stated?
; constant len : (func(2, [(@(0)  @(1)); Int]))

; should the function type be separated?
; (uninterp len (@0 @1) Int)
(uninterp len (Int Int) Int)

; TODO add support for type constructors
; (wf [$k0] ((v [(Tuple Int a)]))))
(wf [$k0] ((v [Silly Int])))

(constraint
(forall ((v [(Tuple Int a)]))
;  (forall ((v [(Silly Int)]))
          (=> (>= (len v) 0) ([$k0]))))

(constraint
; (forall ((v [(Tuple Int a)]))
  (forall ((v [(Silly Int)]))
          (=> ([$k0]) (>= (len v) 0))))
