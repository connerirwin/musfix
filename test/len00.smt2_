(qualif ListZ ((v (Set @0))) (>= (len v) 0))

(define-fun len (@0 @1) Int)

(wf (Set $k0) ((v (Set (Tuple Int @a)))))

(constraint
  (forall ((v (Set (Tuple Int @a))))
    (=> (>= (len v) 0)
        ((Set $k0)))))

(constraint
  (forall ((v (Set (Tuple Int @a))))
    (=> ((Set $k0))
        (>= (len v) 0))))
