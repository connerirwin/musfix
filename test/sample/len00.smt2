(declare-sort Tuple 2)

(declare-fun len ((Set @1)) Int)

(qualif ListZ ((v (Set @0))) (>= (len v) 0))

(wf $k0 ((v (Set (Tuple Int @a)))))

(constraint
  (forall ((v (Set (Tuple Int @a))))
    (=> (>= (len v) 0)
        ($k0 v))))

(constraint
  (forall ((v (Set (Tuple Int @a))))
    (=> ($k0 v)
        (>= (len v) 0))))
