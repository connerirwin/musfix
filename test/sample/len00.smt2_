(declare-sort Tuple 2)

(declare-fun len (@0 @1) Int)

(qualif ListZ ((v (Set @0))) (>= (len v) 0))

(wf (Set $k0) ((v (Set (Tuple Int @a)))))

(constraint
  (forall ((v (Set (Tuple Int @a))))
    (=> (>= (len v) 0)
        ((Set $k0)))))

(constraint
  (forall ((v (Set (Tuple Int @a))))
    (=> ((Set $k0))
        (>= (len v) 0))))
