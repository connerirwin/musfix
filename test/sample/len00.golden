
Preparing to find greatest fixpoint...

Inputs
--------
Reading from file test/sample/len00.msmt
SortDecl "Tuple" 2
UninterpFunction "len" [SetS (VarS "1")] IntS
Qualifier "ListZ" [Var (SetS (VarS "0")) "v"] (Binary Ge (Func IntS "len" [Var (SetS (VarS "1")) "v"]) (IntLit 0))
WFConstraint "$k0" [Var (SetS (DataS "Tuple" [IntS,VarS "a"])) "v"]
HornConstraint [Var (SetS (DataS "Tuple" [IntS,VarS "a"])) "v"] (Binary Implies (Binary Ge (Func IntS "len" [Var (SetS (DataS "Tuple" [IntS,VarS "a"])) "v"]) (IntLit 0)) (Unknown (fromList [("v",Var (SetS (DataS "Tuple" [IntS,VarS "a"])) "v")]) "$k0"))
HornConstraint [Var (SetS (DataS "Tuple" [IntS,VarS "a"])) "v"] (Binary Implies (Unknown (fromList [("v",Var (SetS (DataS "Tuple" [IntS,VarS "a"])) "v")]) "$k0") (Binary Ge (Func IntS "len" [Var (SetS (DataS "Tuple" [IntS,VarS "a"])) "v"]) (IntLit 0)))

QMAP
--------
fromList [("$k0",QSpace {_qualifiers = [], _maxCount = 0})]

Candidates
--------
[]


Final candidates: 

